### ISSUE TITLE
It seems the gradient compuation has bugs. 

### ISSUE THREAD

**Issue by bbsun** [sebgao/cTensor#5]
I am trying following codes. The correct answer should be 6, but it gives 7.  I think it is due to the gradient flow does not propgrate propably. Note here I use the variable xa two times. And that is where is the problem is? If I rename one of them as xb, the result would be correct. Andyway, ctensor is a nice library for studying the autodifferentitation.  
```
import numpy as np
from ctensor import Tensor
from ctensor.functional import conv2d, relu
x  = Tensor(np.array([1]),requires_grad=True)
a  = Tensor(np.array([1]),requires_grad=False)
xa  =   x * a
x2  =   x * x
f_1  =  x * xa
f_2  =  x * xa
f_3  =  x2 * a
f = f_1 +f_3+f_2;
print(f)
f.backward(False)
print(x.grad)
```

**Comment by sebgao** [comment]
Sorry for the late response.

Yes. The gradient backward mechanism has a bug when there exists multiple paths to the final loss. In this case, the divergent node will be backwarded more than once, at each step accumulating gradients to itself and further _backwarding_ this redundant gradients to previous nodes.

This problem can be solved: in the forward process, each node is marked with its outdegree. In the backward process, a node only start its own backward process when it had been visited its outdegree times.

I'll fix this issue ASAP. Thank you for pointing this out!

**Comment by sebgao** [comment]
Maybe solved?
